go/vt/tabletserver/tx_engine.go

```diff
@@ -0,0 +1,252 @@
+// Otherwise, the function waits for all current transactions
+// to conclude.
+func (te *TxEngine) Close(immediate bool) {
+	if !te.isOpen {
+		return
+	}
+	if te.twopcEnabled {
+		te.stopWatchdog()
+		te.readPool.Close()
+	}
+	if immediate {
+		te.rollbackTransactions()
+	}
+	te.txPool.WaitForEmpty()
+	te.txPool.Close()
+	te.isOpen = false
+}
+
+// prepareFromRedo replays and prepares the transactions
+// from the redo log. It also loads previously failed transactions
+// into the reserved list.
+// TODO(sougou): Make this function set the lastId for tx pool to be
+// greater than all those used by dtids. This will prevent dtid
+// collisions.
+func (te *TxEngine) prepareFromRedo() error {
+	ctx := context.Background()
+	var allErr concurrency.AllErrorRecorder
+	conn, err := te.readPool.Get(ctx)
+	if err != nil {
+		return err
+	}
+	defer conn.Recycle()
+	prepared, failed, err := te.twoPC.ReadAllRedo(ctx, conn)
+	if err != nil {
+		return err
+	}
+
+outer:
+	for dtid, tx := range prepared {
+		conn, err := te.txPool.LocalBegin(ctx)
+		if err != nil {
+			allErr.RecordError(err)
+			continue
+		}
+		for _, stmt := range tx {
+			conn.RecordQuery(stmt)
+			_, err := conn.Exec(ctx, stmt, 1, false)
+			if err != nil {
+				allErr.RecordError(err)
+				te.txPool.LocalConclude(ctx, conn)
+				continue outer
+			}
+		}
+		// We should not use the external Prepare because
+		// we don't want to write again to the redo log.
+		err = te.preparedPool.Put(conn, dtid)
+		if err != nil {
+			allErr.RecordError(err)
+			continue
+		}
+	}
+	for _, dtid := range failed {
+		te.preparedPool.SetFailed(dtid)
+	}
+	return allErr.Error()
+}
+
+// rollbackTransactions rolls back all open transactions
+// including the prepared ones.
+// This is used for transitioning from a master to a non-master
+// serving type.
+func (te *TxEngine) rollbackTransactions() {
+	ctx := context.Background()
+	// The order of rollbacks is currently not material because
+	// we don't allow new statements or commits during
+	// this function. In case of any such change, this will
+	// have to be revisited.
+	te.txPool.RollbackNonBusy(ctx)
+	for _, c := range te.preparedPool.FetchAll() {
+		te.txPool.LocalConclude(ctx, c)
+	}
+}
+
+// startWatchdog starts the watchdog goroutine, which looks for abandoned
+// transactions and calls the notifier on them.
+func (te *TxEngine) startWatchdog() {
+	te.ticks.Start(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), te.abandonAge/4)
+		defer cancel()
+		conn, err := te.readPool.Get(ctx)
+		if err != nil {
+			// TODO(sougou): increment error counter.
+			log.Errorf("Error getting connection for 2pc watchdog: %v", err)
+			return
+		}
+		defer conn.Recycle()
+
+		txs, err := te.twoPC.ReadAbandoned(ctx, conn, time.Now().Add(-te.abandonAge))
+		if err != nil {
+			// TODO(sougou): increment error counter.
+			log.Errorf("Error reading transactions for 2pc watchdog: %v", err)
+			return
+		}
+		if len(txs) == 0 {
+			return
```

Comment:
This would spam the logs, given that we'll be polling every few seconds and find nothing most of the time. Instead, I can add a log message in the prepareFromRedo function, which is called once when a tablet becomes master.